File: src/Snaptiq.Core/AlphaNormalizer.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;

namespace Snaptiq.Core;

/// <summary>
/// Converts semi-transparent alpha values to binary transparency while preserving RGB values.
/// </summary>
public static class AlphaNormalizer
{
    /// <summary>
    /// Creates a processed copy of the provided image using the default threshold.
    /// </summary>
    /// <param name="image">The 8-bit RGBA image data to normalize.</param>
    /// <returns>A processed image and processing statistics.</returns>
    public static AlphaNormalizationResult Normalize(Image<Rgba32>? image) =>
        Normalize(image, ThresholdSettings.DefaultThreshold);

    /// <summary>
    /// Creates a processed copy of the provided image using the supplied threshold.
    /// </summary>
    /// <param name="image">The 8-bit RGBA image data to normalize.</param>
    /// <param name="threshold">Alpha threshold in the range 1 through 254.</param>
    /// <returns>A processed image and processing statistics.</returns>
    public static AlphaNormalizationResult Normalize(Image<Rgba32>? image, int threshold)
    {
        ValidateImage(image);
        ThresholdSettings.Validate(threshold);

        Image<Rgba32> processedImage = image.Clone();
        ProcessingStatistics statistics = NormalizeInPlace(processedImage, threshold);

        return new AlphaNormalizationResult(processedImage, statistics);
    }

    /// <summary>
    /// Normalizes alpha values directly on the provided image using the default threshold.
    /// </summary>
    /// <param name="image">The 8-bit RGBA image data to normalize.</param>
    /// <returns>Statistics captured while processing the image.</returns>
    public static ProcessingStatistics NormalizeInPlace(Image<Rgba32>? image) =>
        NormalizeInPlace(image, ThresholdSettings.DefaultThreshold);

    /// <summary>
    /// Normalizes alpha values directly on the provided image using the supplied threshold.
    /// </summary>
    /// <param name="image">The 8-bit RGBA image data to normalize.</param>
    /// <param name="threshold">Alpha threshold in the range 1 through 254.</param>
    /// <returns>Statistics captured while processing the image.</returns>
    public static ProcessingStatistics NormalizeInPlace(Image<Rgba32>? image, int threshold)
    {
        ValidateImage(image);
        ThresholdSettings.Validate(threshold);

        long semiTransparentPixelsFound = 0;
        long pixelsModified = 0;

        image.ProcessPixelRows(accessor =>
        {
            for (int y = 0; y < accessor.Height; y++)
            {
                Span<Rgba32> row = accessor.GetRowSpan(y);

                for (int x = 0; x < row.Length; x++)
                {
                    byte alpha = row[x].A;

                    if (alpha is 0 or 255)
                    {
                        continue;
                    }

                    semiTransparentPixelsFound++;

                    byte normalizedAlpha = alpha < threshold ? (byte)0 : (byte)255;
                    if (normalizedAlpha != alpha)
                    {
                        Rgba32 pixel = row[x];
                        pixel.A = normalizedAlpha;
                        row[x] = pixel;
                        pixelsModified++;
                    }
                }
            }
        });

        return new ProcessingStatistics(
            TotalPixels: (long)image.Width * image.Height,
            SemiTransparentPixelsFound: semiTransparentPixelsFound,
            PixelsModified: pixelsModified,
            ThresholdUsed: threshold);
    }

    private static void ValidateImage(Image<Rgba32>? image)
    {
        if (image is null)
        {
            throw new ArgumentNullException(nameof(image), "Image data cannot be null.");
        }
    }
}

File: web/src/engine/alphaNormalizer.ts
import { validateThreshold } from './thresholdSettings';

export interface ProcessingStatistics {
  totalPixels: number;
  semiTransparentPixelsFound: number;
  pixelsModified: number;
  thresholdUsed: number;
}

export interface AlphaNormalizationResult {
  imageData: ImageData;
  statistics: ProcessingStatistics;
}

export class InvalidImageDataError extends Error {
  public constructor(message: string) {
    super(message);
    this.name = 'InvalidImageDataError';
  }
}

export function normalizeAlpha(imageData: ImageData | null | undefined, threshold: number): AlphaNormalizationResult {
  validateImageData(imageData);
  validateThreshold(threshold);

  const processedData = new Uint8ClampedArray(imageData.data);
  const statistics = normalizeAlphaDataInPlace(processedData, imageData.width, imageData.height, threshold);

  return {
    imageData: new ImageData(processedData, imageData.width, imageData.height),
    statistics
  };
}

export function normalizeAlphaDataInPlace(
  data: Uint8ClampedArray,
  width: number,
  height: number,
  threshold: number
): ProcessingStatistics {
  validateRawRgbaData(data, width, height);
  validateThreshold(threshold);

  let semiTransparentPixelsFound = 0;
  let pixelsModified = 0;

  for (let index = 3; index < data.length; index += 4) {
    const alpha = data[index];

    if (alpha === 0 || alpha === 255) {
      continue;
    }

    semiTransparentPixelsFound += 1;

    const normalizedAlpha = alpha < threshold ? 0 : 255;
    if (normalizedAlpha !== alpha) {
      data[index] = normalizedAlpha;
      pixelsModified += 1;
    }
  }

  return {
    totalPixels: width * height,
    semiTransparentPixelsFound,
    pixelsModified,
    thresholdUsed: threshold
  };
}

function validateImageData(imageData: ImageData | null | undefined): asserts imageData is ImageData {
  if (imageData == null) {
    throw new InvalidImageDataError('Image data cannot be null.');
  }

  validateRawRgbaData(imageData.data, imageData.width, imageData.height);
}

function validateRawRgbaData(data: Uint8ClampedArray, width: number, height: number): void {
  if (!Number.isInteger(width) || width <= 0 || !Number.isInteger(height) || height <= 0) {
    throw new InvalidImageDataError('Image dimensions must be positive integers.');
  }

  if (data.length !== width * height * 4) {
    throw new InvalidImageDataError('RGBA image data length does not match the supplied dimensions.');
  }
}
